File: typescript/packages/core/src/workspace/walk.ts
import type { Stats } from "node:fs";
import { lstat, readdir, readlink } from "node:fs/promises";
import { join } from "node:path";
import { ConfigError } from "../agent/errors";
import { assertNever } from "../assert-never";
import { inRoot } from "../sandbox/tree/paths";
import { denied, excluded } from "./exclude";
import { ignoredPaths, inWorkTree, submodules } from "./git";

// Reads one host directory for a workspace (spec/schema/README.md, Workspace inputs): which paths
// it copies, which it skips (.git, gitignored, deny-listed), and which `include` re-admits.

/** A path the tree takes: a file still on the host, a directory or a symlink. */
export type Found =
  | {
      readonly kind: "file";
      readonly path: string;
      readonly mode: number;
      readonly size: number;
      readonly host: string;
    }
  | { readonly kind: "dir"; readonly path: string; readonly mode: number }
  | {
      readonly kind: "symlink";
      readonly path: string;
      readonly target: string;
    };

/** The limits' running totals, shared by every input (counted after exclusions). */
export type Counted = { files: number; bytes: number };

const MAX_FILES = 10_000;
const MAX_BYTES = 256 * 1024 * 1024;

/**
 * How a directory's children are read: `normal`; `partial`, below a skipped directory, where
 * only paths on the way to an include get through; or `included` from segment `start`, inside an
 * included directory, where only `.git` and deny-listed names below it are skipped.
 */
type State =
  | { readonly kind: "normal" }
  | { readonly kind: "partial" }
  | { readonly kind: "included"; readonly start: number };

type Walk = {
  readonly root: string;
  readonly label: string;
  readonly include: readonly string[];
  readonly ignored: ReadonlySet<string>;
  readonly gitlinks: ReadonlySet<string>;
  readonly used: Set<string>;
  readonly counted: Counted;
  readonly found: Found[];
  readonly skipped: string[];
};

const invalid = (message: string): ConfigError =>
  new ConfigError("invalid_config", message);

const NORMAL: State = { kind: "normal" };
const PARTIAL: State = { kind: "partial" };

const below = (w: Walk, path: string): boolean =>
  w.include.some((i) => i.startsWith(`${path}/`));

/**
 * An `include` entry: its subtree comes in, minus `.git` and deny-listed names below it. The
 * entry counts as used only if something actually skipped this path.
 */
function reAdmit(w: Walk, path: string, state: State): State {
  const skipped =
    state.kind === "partial" ||
    (state.kind === "normal" && (lastDenied(path) || w.ignored.has(path)));
  if (skipped) w.used.add(path);
  return { kind: "included", start: path.split("/").length };
}

/** The state a path is read in, or undefined when it is skipped. */
function decide(w: Walk, path: string, state: State): State | undefined {
  if (w.include.includes(path)) return reAdmit(w, path, state);
  switch (state.kind) {
    case "included":
      return excluded(path, state.start) ? undefined : state;
    case "partial":
      return below(w, path) ? PARTIAL : undefined;
    case "normal":
      if (!lastDenied(path) && !w.ignored.has(path)) return NORMAL;
      return below(w, path) ? PARTIAL : undefined;
    default:
      return assertNever(state);
  }
}

function lastDenied(path: string): boolean {
  const parts = path.split("/");
  return denied(parts, parts.length - 1);
}

/** Throws once the inputs pass a limit. */
export function checkLimits(counted: Counted): void {
  if (counted.files > MAX_FILES)
    throw invalid(
      `workspace: more than ${MAX_FILES} files after exclusions; point localDir at a smaller directory, add a .gitignore, or use the git input`,
    );
  if (counted.bytes > MAX_BYTES)
    throw invalid(
      `workspace: more than 256 MiB of files after exclusions; point localDir at a smaller directory, add a .gitignore, or use the git input`,
    );
}

async function exists(host: string): Promise<boolean> {
  try {
    await lstat(host);
    return true;
  } catch {
    return false;
  }
}

async function nested(w: Walk, path: string, state: State): Promise<void> {
  if (state.kind !== "normal") return;
  if (w.gitlinks.has(path) || (await exists(join(w.root, path, ".git"))))
    throw invalid(
      `${w.label}: ${path} is a git submodule or repository; submodules aren't copied; use the git input or a vendored copy`,
    );
}

async function admit(
  w: Walk,
  path: string,
  st: Stats,
  state: State,
): Promise<void> {
  if (st.isSymbolicLink()) {
    const target = await readlink(join(w.root, path));
    if (!inRoot(path, target))
      throw invalid(
        `${w.label}: symlink ${path} points outside it (${target})`,
      );
    w.found.push({ kind: "symlink", path, target });
  } else if (st.isDirectory()) {
    await nested(w, path, state);
    w.found.push({ kind: "dir", path, mode: 0o755 });
    await walkDir(w, path, state);
  } else if (st.isFile()) {
    const mode = (st.mode & 0o111) === 0 ? 0o644 : 0o755;
    w.found.push({
      kind: "file",
      path,
      mode,
      size: st.size,
      host: join(w.root, path),
    });
    w.counted.files += 1;
    w.counted.bytes += st.size;
    checkLimits(w.counted);
  } else
    throw invalid(
      `${w.label}: ${path} is not a regular file, directory or symlink`,
    );
}

async function walkDir(w: Walk, dir: string, state: State): Promise<void> {
  const names = (await readdir(join(w.root, dir))).toSorted();
  for (const name of names) {
    const path = dir === "" ? name : `${dir}/${name}`;
    const next = name === ".git" ? undefined : decide(w, path, state);
    if (next === undefined) w.skipped.push(path);
    else await admit(w, path, await lstat(join(w.root, path)), next);
  }
}

/** What reading one directory gives: its paths, and the top-most ones it skipped, sorted. */
export type Read = {
  readonly found: readonly Found[];
  readonly skipped: readonly string[];
};

/**
 * Reads `root` (an existing host directory) as a workspace source. `used` collects the include
 * entries that re-admitted a skipped path; `counted` carries the limits across inputs.
 */
export async function readDir(
  root: string,
  label: string,
  include: readonly string[],
  used: Set<string>,
  counted: Counted,
): Promise<Read> {
  const git = await inWorkTree(root, label);
  const w: Walk = {
    root,
    label,
    include,
    ignored: git ? await ignoredPaths(root, label) : new Set(),
    gitlinks: git ? await submodules(root, label) : new Set(),
    used,
    counted,
    found: [],
    skipped: [],
  };
  await walkDir(w, "", NORMAL);
  return { found: w.found, skipped: w.skipped.toSorted() };
}

File: python/src/threads/workspace/walk.py
"""Reads one host directory for a workspace (spec/schema/README.md, Workspace inputs): which
paths it copies, which it skips (.git, gitignored, deny-listed), and which `include` re-admits."""

import os
from collections.abc import Sequence
from dataclasses import dataclass, field
from pathlib import Path
from typing import Final, Literal, assert_never

from threads.agents.config import ConfigError
from threads.log.jcs import utf16_key
from threads.sandbox.tree.paths import in_root
from threads.workspace.exclude import denied, excluded
from threads.workspace.git import ignored_paths, in_work_tree, invalid, submodules

MAX_FILES: Final = 10_000
MAX_BYTES: Final = 256 * 1024 * 1024


@dataclass(frozen=True, slots=True)
class FoundFile:
    path: str
    mode: int
    size: int
    host: Path


@dataclass(frozen=True, slots=True)
class FoundDir:
    path: str
    mode: int = 0o755


@dataclass(frozen=True, slots=True)
class FoundSymlink:
    path: str
    target: str


type Found = FoundFile | FoundDir | FoundSymlink


@dataclass(slots=True)
class Counted:
    """The limits' running totals, shared by every input (counted after exclusions)."""

    files: int = 0
    nbytes: int = 0


type State = Literal["normal", "partial"] | int
"""How a directory's children are read: `normal`; `partial`, below a skipped directory, where
only paths on the way to an `include` get through; or, inside an included directory, the segment
count it started at, where only `.git` and deny-listed names below it are skipped."""


@dataclass(slots=True)
class Walk:
    root: Path
    label: str
    include: Sequence[str]
    ignored: frozenset[str]
    gitlinks: frozenset[str]
    used: set[str]
    counted: Counted
    found: list[Found] = field(default_factory=list[Found])
    skipped: list[str] = field(default_factory=list[str])


def check_limits(counted: Counted) -> None:
    """Raises once the inputs pass a limit."""
    advice = "point local_dir at a smaller directory, add a .gitignore, or use the git input"
    if counted.files > MAX_FILES:
        raise invalid(f"workspace: more than {MAX_FILES} files after exclusions; {advice}")
    if counted.nbytes > MAX_BYTES:
        raise invalid(f"workspace: more than 256 MiB of files after exclusions; {advice}")


def _last_denied(path: str) -> bool:
    parts = path.split("/")
    return denied(parts, len(parts) - 1)


def _below(w: Walk, path: str) -> bool:
    return any(i.startswith(f"{path}/") for i in w.include)


def _re_admit(w: Walk, path: str, state: State) -> State:
    """An `include` entry: its subtree comes in, minus `.git` and deny-listed names below it.
    The entry counts as used only if something actually skipped this path."""
    was_skipped = state == "partial" or (
        state == "normal" and (_last_denied(path) or path in w.ignored)
    )
    if was_skipped:
        w.used.add(path)
    return len(path.split("/"))


def _decide(w: Walk, path: str, state: State) -> State | None:
    """The state a path is read in, or None when it is skipped."""
    if path in w.include:
        return _re_admit(w, path, state)
    match state:
        case int():
            return None if excluded(path, state) else state
        case "partial":
            return "partial" if _below(w, path) else None
        case "normal":
            if not _last_denied(path) and path not in w.ignored:
                return "normal"
            return "partial" if _below(w, path) else None
        case _:
            assert_never(state)


def _nested(w: Walk, path: str, state: State) -> None:
    if state != "normal":
        return
    if path in w.gitlinks or (w.root / path / ".git").exists():
        raise invalid(
            f"{w.label}: {path} is a git submodule or repository; submodules aren't copied; "
            "use the git input or a vendored copy"
        )


def _admit(w: Walk, path: str, entry: os.DirEntry[str], state: State) -> None:
    host = w.root / path
    if entry.is_symlink():
        target = os.readlink(host)
        if not in_root(path, target):
            raise invalid(f"{w.label}: symlink {path} points outside it ({target})")
        w.found.append(FoundSymlink(path, target))
        return
    if entry.is_dir():
        _nested(w, path, state)
        w.found.append(FoundDir(path))
        _walk(w, path, state)
        return
    if not entry.is_file():
        raise invalid(f"{w.label}: {path} is not a regular file, directory or symlink")
    stat = entry.stat(follow_symlinks=False)
    mode = 0o644 if stat.st_mode & 0o111 == 0 else 0o755
    w.found.append(FoundFile(path, mode, stat.st_size, host))
    w.counted.files += 1
    w.counted.nbytes += stat.st_size
    check_limits(w.counted)


def _walk(w: Walk, directory: str, state: State) -> None:
    entries = sorted(os.scandir(w.root / directory), key=lambda e: e.name)
    for entry in entries:
        path = entry.name if directory == "" else f"{directory}/{entry.name}"
        nxt = None if entry.name == ".git" else _decide(w, path, state)
        if nxt is None:
            w.skipped.append(path)
        else:
            _admit(w, path, entry, nxt)


@dataclass(frozen=True, slots=True)
class Read:
    """What reading one directory gives: its paths, and the top-most ones it skipped, sorted."""

    found: tuple[Found, ...]
    skipped: tuple[str, ...]


async def read_dir(
    root: Path, label: str, include: Sequence[str], used: set[str], counted: Counted
) -> Read:
    """Reads `root` (an existing host directory) as a workspace source. `used` collects the
    include entries that re-admitted a skipped path; `counted` carries the limits across
    inputs."""
    tracked = await in_work_tree(root, label)
    w = Walk(
        root=root,
        label=label,
        include=include,
        ignored=await ignored_paths(root, label) if tracked else frozenset(),
        gitlinks=await submodules(root, label) if tracked else frozenset(),
        used=used,
        counted=counted,
    )
    _walk(w, "", "normal")
    # Sorted by UTF-16 code units, like every path list on the wire.
    return Read(tuple(w.found), tuple(sorted(w.skipped, key=utf16_key)))


__all__ = [
    "ConfigError",
    "Counted",
    "Found",
    "FoundDir",
    "FoundFile",
    "FoundSymlink",
    "Read",
    "check_limits",
    "read_dir",
]
